src/main/java/com/microsoft/azure/proton/transport/proxy/impl/ProxyImpl.java

```diff
@@ -274,5 +307,104 @@ public void close_head() {
             headClosed = true;
             underlyingOutput.close_head();
         }
+
+        private void getChallengeQuestionHeaders(String line, Map<String, String> challengeQuestionValues) {
+            String context =  line.substring(PROXY_AUTH_DIGEST.length());
+            String[] headerValues = context.split(",");
+
+            for (String headerValue : headerValues) {
+                if (headerValue.contains("=")) {
+                    String key = headerValue.substring(0, headerValue.indexOf("="));
+                    String value = headerValue.substring(headerValue.indexOf("=") + 1);
+                    challengeQuestionValues.put(key.trim(), value.replaceAll("\"", "").trim());
+                }
+            }
+        }
+
+        private void computeBasicAuthHeader(){
+            final PasswordAuthentication authentication = Authenticator.requestPasswordAuthentication(
+                    "",
+                    null,
+                    0,
+                    "https",
+                    "Event Hubs client websocket proxy support",
+                    "basic",
+                    null,
+                    Authenticator.RequestorType.PROXY);
+            if (authentication == null) return;
+
+            final String proxyUserName = authentication.getUserName();

```

Comment:
1. Please remove usage of Authenticator - in this library. We want to support even if there are other ways for java apps to specify the proxy's username and password.
2. Introduce new constructor for ProxyImpl to accept username and password. 
3. Indicate on the javadoc comments for ProxyImpl.configure() - header parametet - to explicitly tell that - Authorization header should not be added. Add a check to throw if it is added.

Reply:
Different auth methods can have different credentials. We need to be able to interrogate Authenticator for creds on a per-method basis, starting with the most preferred method and falling back until we find one that has creds. Passing a single set of creds to ProxyImpl is not sufficient.

Considering that Java's own HTTP auth support is built around Authenticator, I don't see a reason to not take the same dependency.

Reply:
UserName and Password are the common denominator that this class needed for both schemes - basicAuth and digest Auth, i.e., these are what are being queried out of Authenticator. Do you think, there is more info needed?
Pre java 1.5 users are used to specify the auth parameters on command line. But, after java 1.5 Authenticator came into picture. We took this approach, so that we can support both the cases.

If from this version on, we want to take the approach of directly calling Authenticator API's in proton-j-extensions library itself - lets also bring ProxySelector to this layer - to keep the Proxy story similar for getting all Proxy connection detail - i.e., ip and auth details..

---
In reply to: [260928788](https://github.com/Azure/qpid-proton-j-extensions/pull/10#discussion_r260928788) [](ancestors = 260928788)

Reply:
I take the second approach that supporting ProxySelector and Authenticator. As I know, proton-J is using java 8.

Reply:
Authenticator has existed since Java 1.2. Can we get some clarity over what is being requested here so that Shawn may make progress here. Thanks.

Reply:
Shawn and I discussed this in IM. Proton-J requires Java 1.8 so pre-1.5 is irrelevant as you note. We concluded that keeping Authenticator and moving ProxySelector was the way to go. I will review the latest commits on this PR today.

Reply:
Hi @JonathanGiles! Thanks a lot for pitching in! 

Today this library that @mssfang is contributing to - is used by `EventHubs`, `ServiceBus` and `IotHub` libraries. The existing way those teams consume this library so far - is by figuring out Proxy related parameters (like the info you would get out of `ProxySelector` and `Authenticator`) in their layer and passing them onto this library.
 With the current change @mssfang intends to do - it would change and move extracting proxy related info directly in `proton-j-extensions` layer; In the process he only got `Authenticator` to this layer. We were requesting him to get the `ProxySelector` to this layer as well - so that all proxy related parameters move here.

The mention about the versions - was not about the JDK compatibility. It is precisely about Customer Habits :). Some of the customers who were dogfooding this feature for EventHubs were trying to set the Proxy parameters via java command line. So, back-in-time, we thought we will leave `proton-extensions` library away from the logic of extracting Proxy parameters diff sources.

Does this make sense? Please feel free to schedule a meeting - if you folks need more inputs... 
Sorry about the delay in responses - I moved out from `Event Hubs` to `Azure Data` and struggling to find time for this.

Reply:
Thank you for your reply. Current PR has the newest changes including the ProxySelector in the proton-j-extension layer. Let me know if there is anything missing or not satisfying the requirement.